### ISSUE TITLE
Question for the metric 

### ISSUE THREAD

**Issue by SCUTykLin** [Kunhao-Liu/StyleGaussian#5]
In your paper, u say:
"For multi-view consistency, following previous work [7, 8, 15, 24, 29], we warp one view to another according to the optical flow [44] using softmax splatting [33], and then compute the masked RMSE score
and LPIPS score [55] to measure stylization consistency. "

Can u provide the code? Thanks

**Comment by Kunhao-Liu** [comment]
Hi, we have modified the `demo.py` file in [RAFT](https://github.com/princeton-vl/RAFT) to calculate the score. Here's the code:

```python
import sys
sys.path.append('core')

import argparse
import os
import cv2
import imageio
import glob
import numpy as np
import torch
import torch.nn.functional as F
from PIL import Image

from raft import RAFT
from utils import flow_viz
from utils.utils import InputPadder

import torchvision
import softsplat
import lpips

##########################################################

backwarp_tenGrid = {}

def backwarp(tenIn, tenFlow):
    if str(tenFlow.shape) not in backwarp_tenGrid:
        tenHor = torch.linspace(-1.0 + (1.0 / tenFlow.shape[3]), 1.0 - (1.0 / tenFlow.shape[3]), tenFlow.shape[3]).view(1, 1, 1, -1).repeat(1, 1, tenFlow.shape[2], 1)
        tenVer = torch.linspace(-1.0 + (1.0 / tenFlow.shape[2]), 1.0 - (1.0 / tenFlow.shape[2]), tenFlow.shape[2]).view(1, 1, -1, 1).repeat(1, 1, 1, tenFlow.shape[3])

        backwarp_tenGrid[str(tenFlow.shape)] = torch.cat([tenHor, tenVer], 1).cuda()
    # end

    tenFlow = torch.cat([tenFlow[:, 0:1, :, :] / ((tenIn.shape[3] - 1.0) / 2.0), tenFlow[:, 1:2, :, :] / ((tenIn.shape[2] - 1.0) / 2.0)], 1)

    return torch.nn.functional.grid_sample(input=tenIn, grid=(backwarp_tenGrid[str(tenFlow.shape)] + tenFlow).permute(0, 2, 3, 1), mode='bilinear', padding_mode='zeros', align_corners=False)
# end

##########################################################


DEVICE = 'cuda'

def load_image(imfile):
    img = np.array(Image.open(imfile)).astype(np.uint8)
    img = torch.from_numpy(img).permute(2, 0, 1).float()
    return img[None].to(DEVICE)

def warp(x, flo, padding_mode='border'):
    B, C, H, W = x.size()

    # Mesh grid
    xx = torch.arange(0, W).view(1,-1).repeat(H,1)
    yy = torch.arange(0, H).view(-1,1).repeat(1,W)
    xx = xx.view(1,1,H,W).repeat(B,1,1,1)
    yy = yy.view(1,1,H,W).repeat(B,1,1,1)
    grid = torch.cat((xx,yy),1).float()

    if x.is_cuda:
        grid = grid.cuda()
    vgrid = grid - flo
    
    # Scale grid to [-1,1]
    vgrid[:,0,:,:] = 2.0*vgrid[:,0,:,:]/max(W-1,1)-1.0
    vgrid[:,1,:,:] = 2.0*vgrid[:,1,:,:]/max(H-1,1)-1.0
    vgrid = vgrid.permute(0,2,3,1)
    output = F.grid_sample(x, vgrid, padding_mode=padding_mode, mode='nearest')
    return output

def viz(img, flo):
    img = img[0].permute(1,2,0).cpu().numpy()
    flo = flo[0].permute(1,2,0).cpu().numpy()
    
    # map flow to rgb image
    flo = flow_viz.flow_to_image(flo)
    # img_flo = np.concatenate([img, flo], axis=0)

    # import matplotlib.pyplot as plt
    # plt.imshow(img_flo / 255.0)
    # plt.show()

    # cv2.imshow('image', img_flo[:, :, [2,1,0]]/255.0)
    # cv2.waitKey()

    imageio.imwrite('./vis.png', flo)



def demo(args):
    model = torch.nn.DataParallel(RAFT(args))
    model.load_state_dict(torch.load(args.model))

    model = model.module
    model.to(DEVICE)
    model.eval()

    loss_fn_vgg = lpips.LPIPS(net='vgg').cuda()

    with torch.no_grad():
        # images = glob.glob(os.path.join(args.path, '*.png')) + \
        #          glob.glob(os.path.join(args.path, '*.jpg'))
        
        # images = sorted(images)
        # imfile1 = images[0]
        # imfile2 = images[10]

        image1 = load_image(args.img1)
        image2 = load_image(args.img2)

        padder = InputPadder(image1.shape)
        image1, image2 = padder.pad(image1, image2)

        flow_low, flow_up = model(image1, image2, iters=32, test_mode=True)
        
        # torchvision.utils.save_image(flow_low, './flow_low.png')
        # print(flow_low.size()) #torch.Size([1, 2, 95, 126])
        # torchvision.utils.save_image(flow_up, './flow_up.png')

        # viz(image1, flow_up)

        # warped = warp(image1, flow_up)
        # warped = warped[0].permute(1,2,0).cpu().numpy()
        # imageio.imwrite('./warped.png', warped)

        tenMetric = torch.nn.functional.l1_loss(input=image1, target=backwarp(tenIn=image2, tenFlow=flow_up), reduction='none').mean([1], True)
        alpha = 10
        tenSoftmax = softsplat.softsplat(tenIn=image1, tenFlow=flow_up, tenMetric=(-alpha * tenMetric).clip(-alpha, alpha), strMode='soft') 

        # imageio.imwrite(f'./softmax_{alpha}.png', tenSoftmax[0, :, :, :].permute(1,2,0).cpu().numpy())

        mask =  torch.where(
                tenSoftmax.sum(1, keepdim=True) == 0.,
                0.,
                1.
            )

        lpips_score = loss_fn_vgg(tenSoftmax,image2*mask)


        rmse = torch.sqrt(torch.nn.functional.mse_loss(tenSoftmax/255.,image2*mask/255.,reduction='mean'))

        print(f"lpips:{lpips_score.item()},rmse:{rmse}")



if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('--model', default='/data/Projects/RAFT/models/raft-things.pth', help="restore checkpoint")
    parser.add_argument('--path', help="dataset for evaluation")
    parser.add_argument('--small', action='store_true', help='use small model')
    parser.add_argument('--mixed_precision', action='store_true', help='use mixed precision')
    parser.add_argument('--alternate_corr', action='store_true', help='use efficent correlation implementation')
    parser.add_argument('--img1', default='/mnt/sda/Projects/Posefree_Dynamic_NeRF/dataset/DAVIS_1080p/horsejump-high/images/00000.jpg')
    parser.add_argument('--img2', default='/mnt/sda/Projects/Posefree_Dynamic_NeRF/dataset/DAVIS_1080p/horsejump-high/images/00001.jpg')
    args = parser.parse_args()

    demo(args)

```

**Comment by Lans1ot** [comment]
Thank you for kindly providing the code for evaluating LPIPS and RMSE.
